### ISSUE TITLE
Undefined symbols when trying to access properties in annotation instances

### ISSUE THREAD

**Issue by soywiz** [JetBrains/kotlin-native#1725]
In an MPP project I had some `annotation class MyAnnotation(val myfield: Int)`, and then one unreferenced function in common `fun myfunc(annotation: MyAnnotation): Int { return annotation.myfield }`. In fact I was able to get the annotation instance in the JVM but that code was in common, and was valid, and produced no compilation errors in common.

In my specific case I had this error at linking time:

```
annotation class FontDescriptor(val face: String, val size: Int, val chars: String = "0123456789")
```

```
Task :sample1-native:compileDebugKotlinNative
Undefined symbols for architecture x86_64:
  "_kfun:com.soywiz.korge.bitmapfont.FontDescriptor.<get-chars>()kotlin.String", referenced from:
      _kfun:com.soywiz.korge.bitmapfont.BitmapFontAsyncFactory.$create$1.doResume(kotlin.Any?;kotlin.Throwable?)kotlin.Any? in combined.o
  "_kfun:com.soywiz.korge.bitmapfont.FontDescriptor.<get-face>()kotlin.String", referenced from:
      _kfun:com.soywiz.korge.bitmapfont.BitmapFontAsyncFactory.$create$1.doResume(kotlin.Any?;kotlin.Throwable?)kotlin.Any? in combined.o
  "_kfun:com.soywiz.korge.bitmapfont.FontDescriptor.<get-size>()ValueType", referenced from:
      _kfun:com.soywiz.korge.bitmapfont.BitmapFontAsyncFactory.$create$1.doResume(kotlin.Any?;kotlin.Throwable?)kotlin.Any? in combined.o
  "_kfun:com.soywiz.korge.resources.Path.<get-path>()kotlin.String", referenced from:
      _kfun:com.soywiz.korge.atlas.Atlas.Factory.$create$1.doResume(kotlin.Any?;kotlin.Throwable?)kotlin.Any? in combined.o
      _kfun:com.soywiz.korge.resources.ResourcesRoot.get(com.soywiz.korge.resources.Path)com.soywiz.korio.file.VfsFile in combined.o
      _kfun:com.soywiz.korge.audio.SoundFile.Factory.$create$5.doResume(kotlin.Any?;kotlin.Throwable?)kotlin.Any? in combined.o
ld: symbol(s) not found for architecture x86_64
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld invocation reported errors
```

Maybe you can generate stubs for this case (throwing an exception for example)? Or disallow accessing annotation properties in common.

If you cannot generate an instance of an annotation in native, probably those methods would be unreferenced and could be stripped out by treeshaking, so the stubs wouldn't be included, but it wouldn't fail either if no treeshaking is performed.

**Comment by olonho** [comment]
General issue discussed in https://youtrack.jetbrains.com/issue/KT-25108

**Comment by olonho** [comment]
What is the point of having that code in common?

**Comment by soywiz** [comment]
No much point right now. The story:

Before using MPP for those libraries, I was using Kotlin JVM, and then targeting other places using an AOT converting JVM bytecode into C++ and JS source-code.

With that AOT I had full JVM reflection capabilities, and I did Dependency Injection and was passing some annotationes to factories to build objects using that information.

Once Kotlin started to support MPP, I started migrating it to JVM and JS.
Now that kotlin-native supports gradle and it is more mature, I have started to target Native too.
And happened that there is still legacy code there. Right now I'm doing dependency injection by manually mapping classes and in future versions I guess I will totally remove the dependency injection for the first version.

Not sure if in the future using a kotlin compiler plugin we will be able to do proper DI with annotation support, but right now it is not the case.

Still I found the issue and wanted to report it.

My only concern at this point is that this is a linking-time error. I would prefer no error at all, but if an error is required I would prefer for all MPP/common, or at least a compile-time error at native compilation with a descriptive message.
